network/phonebook.go

```diff
@@ -104,48 +132,62 @@ func (e *phonebookEntries) filterRetryTime(t time.Time) []string {
 // new entries in they are being added
 // existing items that aren't included in they are being removed
 // matching entries don't change
-func (e *phonebookEntries) ReplacePeerList(they []string) {
+func (e *phonebookImpl) ReplacePeerList(dnsAddresses []string, networkName string) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
 
 	// prepare a map of items we'd like to remove.
 	removeItems := make(map[string]bool, 0)
 	for k := range e.data {
 		removeItems[k] = true
 	}
 
-	for _, addr := range they {
-		if _, has := e.data[addr]; has {
-			// we already have this. do nothing.
+	for _, addr := range dnsAddresses {
+		if pbData, has := e.data[addr]; has {
+			// we already have this.
+			// Update the networkName
+			pbData.networkNames[networkName] = true
+
+			// do nor remove this entry
 			delete(removeItems, addr)
 		} else {
 			// we don't have this item. add it.
-			e.data[addr] = phonebookData{}
+			e.data[addr] = makePhonebookData(networkName)
 		}
 	}
 
-	// remove items that were missing in they
+	// remove items that were missing in dnsAddresses
 	for k := range removeItems {
-		delete(e.data, k)
+		e.deletePhonebookEntry(k, networkName)
```

Comment:
Instead of this, shouldn't we avoid placing entries in `removeItems` that shouldn't be deleted ?
i.e. entries in e.data that aren't in networkNames[] shouldn't be added to start with.